infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/RemoveLocalProjectsFolderOnWorkspaceRemove.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.docker.local.projects;
+
+import static java.lang.System.getenv;
+import static org.eclipse.che.commons.lang.IoUtil.deleteRecursive;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import com.google.inject.name.Named;
+import java.io.File;
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import org.eclipse.che.api.core.model.workspace.Workspace;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.core.notification.EventSubscriber;
+import org.eclipse.che.api.workspace.server.event.WorkspaceRemovedEvent;
+import org.eclipse.che.commons.lang.concurrent.LoggingUncaughtExceptionHandler;
+import org.eclipse.che.commons.lang.concurrent.ThreadLocalPropagateContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Listener for remove workspace files after {@code WorkspaceRemovedEvent}.
+ *
+ * @author Alexander Andrienko
+ * @author Sergii Leshchenko
+ */
+@Singleton
+public class RemoveLocalProjectsFolderOnWorkspaceRemove
+    implements EventSubscriber<WorkspaceRemovedEvent> {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(RemoveLocalProjectsFolderOnWorkspaceRemove.class);
+
+  private final ExecutorService executor;
+  private final LocalProjectsFolderPathProvider workspaceFolderPathProvider;
+
+  @Inject(optional = true)
+  @Named("host.projects.root")
+  private String hostProjectsFolder;
+
+  @Inject
+  public RemoveLocalProjectsFolderOnWorkspaceRemove(
+      LocalProjectsFolderPathProvider workspaceFolderPathProvider) {
+    this.workspaceFolderPathProvider = workspaceFolderPathProvider;
+    executor =
+        Executors.newCachedThreadPool(
+            new ThreadFactoryBuilder()
+                .setNameFormat("RemoveLocalProjectsFolderOnWorkspaceRemove-%d")
+                .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+                .setDaemon(true)
```

Comment:
In this case, we may have untraced unremoved folders that litter the system. Should we use non-daemon threads instead?

Reply:
It was implemented in this way. So I think It makes sense to change it in way you proposed